dopamine/discrete_domains/run_experiment.py

```diff
@@ -582,8 +590,9 @@ def __init__(
     """
     threading_utils.initialize_local_attributes(
         self, _environment=create_environment_fn)
-    self._running_iterations = threading.Semaphore(max_simultaneous_iterations)
     self._output_lock = threading.Lock()
+    self._running_iterations = queue.Queue(max_simultaneous_iterations)
```

Comment:
I think `max_simultaneous_iterations` is being put to double purpose.  It's limiting the size of each queue.  I think I understand how experience gathering is being limited by the queue size.  But I'm not following how training is.  Isn't there quest one single training worker?

Reply:
There is one worker for training, that's right.

This queue accumulates the number of training steps to complete. The goal is to have training steps occur simultaneously to experience generation. Not faster or not slower (you don't want all training steps to run at the end of experience generation).

The goal of this limit is to prevent training from accumulating delay compared to experience generation. If the training queue is full, then experience generation is too fast and is going to be slowed down.
On the opposite, training steps to run are added to the queue only when a experience step has completed.

Reply:
Some comments and requests for clarification up in the README.  To the lines of code at hand, why choose `max_simultaneous_iterations` as the size of the training queue?

Reply:
Trying to answer my own question: an iteration is defined as a pair of experience-gathering step and training step.  Two iterations *(e_i, t_i)* and *(e_j, t_j)* are simultaneous if either of *e_i* or *t_i* are enqueued at the same time as either of *e_j* or *t_j*.  Setting the the size of the training queue to `max_simultaneous_iterations` (= *s*) keeps no more than *s* training steps simultaneous.  There are then further limitations that ensure no other *e_i*s are enqueued other than those paired with training steps in the queue?

Reply:
I think several values might work. `max_simultaneous_iterations` is at the limit and allow to handle the case where all thread complete at the same time, hence adding `max_simultaneous_iterations` items to the queue, without blocking. If `max_simultaneous_iterations + 1` are being added, it means experience generation is faster than training so in this regard `max_simultaneous_iterations` seemed a right choice.